### ISSUE TITLE
statistic functions can not accept single expression as input

### ISSUE THREAD

**Issue by judovana** [gbenroscience/ParserNG#25]
Hello!
This is very severe issue which is prohibiting usage of parserng in deployment.

If statistic function is given eg `sum(5,4)` it works, when it is given `sum(5)` it works. 
When it is given `sum(1+1,2+2)` it works. However when it is given `sum(1+1)` it fails.
Notable is, that eg `sum(sin(5))` works too. However `sum(sin(5)+(sin(5))` don't. 

All my inputs for parserNG are generated, so it took some time to see why some of the *first* calls fails. Yes first, as in in first call they have just `one` argument. Also not all equations have expression in first argument.. 

The issue is reproduced here: https://github.com/gbenroscience/ParserNG/pull/23
Possible workaround suggested at: https://github.com/gbenroscience/ParserNG/pull/24
The issue is not related to https://github.com/gbenroscience/ParserNG/pull/22 ;  was jsut accidently hit there.

The fix would be harder. It seems `coma` is somewhere handled wrongly. I tried, but but failed. IMHO,  parserNG do nto have enough juint coverage to actually affrod proepr fix. But you know internals better.

The internal isse is visibel like this:
`sum(1,2)` is proeperly parsed and sent to `sum` impelmetnation as two arguments   as `1,2` 
`sum(1+1,2+2)` is proeperly parsed, evaluated and sent to `sum` impelmetnation as two arguments  as `2,4`
however `sum(1+1)` is wronlgy parsed and not evaluated, and send to `sum` implemetnation as three arguments `1,+,1`

@gbenroscience  @gbenrojiboye   can you fix it?  If not, can you please participate on workaround? On short run, the workaround is just: https://github.com/gbenroscience/ParserNG/pull/24/files#diff-7592ec3fb1f62c29d86d7c8c826fc5a684c85caee27fc35b7b6c3c5628bc47e1R101
Which can be then reused in any existing statistic function, or in any override of statistic function (see https://github.com/gbenroscience/ParserNG/pull/24/commits/0c7177e85e60f6e7ba5eeb88a3daa9dd988ad67f)

The workaround is checking the input on function. If it is all numbers, then it is passed in.
If it is not all numbers, then it is joined first and evaluated. Then result is passed in.

Unless the issue is resolved "properly" all the hardcoded statistic function can reuse above workaround.

There is acompanying issue, when hardcoded not-statistic functions  eg sin or cos  can actually take several argument, and *missinterpret* them:
`sin(5)` ok. `sin(5+5`) ok. However `sin(5,5)` is accepted, and  evaluated as `sin((5)(5))` whis is `sin(5*5)` which is wrong.
This issue is not fatal, and as blocking as first issue, 
Still the fix for both issues seems to go deep into parsetr. On longest run the statistic an non statistc functions should be aligned.

The issue is also reproduced here: https://github.com/gbenroscience/ParserNG/pull/23
and also workaround suggested at: https://github.com/gbenroscience/ParserNG/pull/24 by moving non statistic sin to statistic, and properly throwing exception if more then one argument is set.

@gbenrojiboye  @gbenroscience  Please, look into the PRs or suggest better solution, otherwise we will need to live only in the fork of ParserNG instead of  contributing back.

Important highlight: @gbenrojiboye  @gbenroscience  this issue **is not caused by any recent commits.** I had tracked it down to **first** public commit on this repo

**Comment by gbenroscience** [comment]
Hi @judovana. Happy New Year😁
Sorry I just saw this today and I am really busy at the moment.
I will check the parser innards and fix. Will let you know when it's done.
Thanks

**Comment by judovana** [comment]
Hello! Welcome! All the best to you too!
I'm really glad to see you back. If you think it can be fixed in the main parser logic, then I can happily drop the https://github.com/gbenroscience/ParserNG/pull/24. Similarly, it  is bad idea to merge https://github.com/gbenroscience/ParserNG/pull/23, but that should be part of your's fixing change.

I will close https://github.com/gbenroscience/ParserNG/pull/24 ; Feel free to close https://github.com/gbenroscience/ParserNG/pull/23 but I think it should remain as remainder, and eg merged after you are done with the fix or with it..

Unless you are sure you can fix this in week or two, which I'm afraid you can not, as you are very busy man. Are you ok with me, adding PR, which will allow to workaround it in consuming projects, and finish rounding work?
The workaround can be removed once you are done with the fix, but currently inability to proceed with  https://github.com/gbenroscience/ParserNG/pull/22 and sum/prod and others crashing with one argument, are really really painful.

**Comment by gbenroscience** [comment]
Yes, go ahead with the workaround while I implement the fix.
Thanks

**Comment by judovana** [comment]
Workaroudn is now  https://github.com/gbenroscience/ParserNG/pull/26 is now ok to merge (if it is ok by you).

**Comment by judovana** [comment]
> ahead with the workaround while I implement the fix.
> Thanks

You just closed 26. Unless you have fix ready, that was not nice:(

**Comment by judovana** [comment]
And even if you have fix. The extracting of methods is good thing. The superlong switches are not nice.

**Comment by gbenroscience** [comment]
> > ahead with the workaround while I implement the fix.
> > Thanks
> 
> You jsut closed 26. Unless you have fix ready, that was not nice:(

Sorry, maybe I misunderstood you. Please reopen it.

**Comment by gbenroscience** [comment]
I have reopened it.

**Comment by gbenroscience** [comment]
Hi @judovana , so I had some time yesterday and I have fixed the issue as I promised.
I will be removing the workaround and force pushing the fix.
I assume this works for you?

**Comment by gbenroscience** [comment]
Hope you respond asap as I have quite a schedule, lol

**Comment by judovana** [comment]
Hello! Sure!

First of all, thanx a lot for merging the workaround. I was just rebasing my other three PRs, so they can make advantage of the workaround, and then reply here. But if yo-u have fix, I rather stop rebasing, and am replying:)
I will finish the rebase of https://github.com/gbenroscience/ParserNG/pull/23 as it may be helpful to your fix.

If you have fix, then congratulations. Am looking forwartd for PR/commit to see that. The only thing you shoudl do to get rid of the workaround, is
 - repalce all usages of Utils.evaluateSingleToken  by Utils.tokensToNumbers
xor
 - replace body of Utils.evaluateSingleToken() simply by calling tokensToNumbers

I vote for the first opinion - to repalce all usages of Utils.evaluateSingleToken  by Utils.tokensToNumbers

Does your fix improve also handling of sin(1,2) being sin(1*2)?

//me speeding up the rebase of the test

**Comment by judovana** [comment]
Hello! https://github.com/gbenroscience/ParserNG/pull/23 is now rebased. The ProdTest have one failing test, which should be fixed by yoour's fix.
SinTest have four failures. Yet again, they shoudl be fixed if your fix is fixing both issues. Will now rebase also toher PRs, thoug they are not directly relevant.

**Comment by gbenroscience** [comment]
Sorry, bad practise. I already merged with master. Can you pull the master and do your checks?

**Comment by gbenroscience** [comment]
**_Does your fix improve also handling of sin(1,2) being sin(1*2)?_**
I was not able to reproduce this issue on my original builds, could it be from one of your commits? let me know so we can track it down speedily. Thanks

**Comment by gbenroscience** [comment]
Was the **_sin_** overloaded or something like that? lol?

**Comment by judovana** [comment]
No sin was never overloaded, and is not going to be, unless necessary. The closed  https://github.com/gbenroscience/ParserNG/pull/24  was jsut PoC

You are right.
```
wget https://repo1.maven.org/maven2/com/github/gbenroscience/parser-ng/0.1.5/parser-ng-0.1.5.jar
java -cp  parser-ng-0.1.5.jar   parser.cmd.ParserCmd
wget https://repo1.maven.org/maven2/com/github/gbenroscience/parser-ng/0.1.6/parser-ng-0.1.6.jar
java -cp  parser-ng-0.1.6.jar   parser.cmd.ParserCmd
wget https://repo1.maven.org/maven2/com/github/gbenroscience/parser-ng/0.1.8/parser-ng-0.1.8.jar
java -cp  parser-ng-0.1.8.jar   parser.cmd.ParserCmd
```
All behaves correctly and sin(1,1) is  evaluated as `A SYNTAX ERROR OCCURRED` in the `parser.cmd.ParserCmd`
Weird.... You see, also latest version where all is in:
```
 java -jar target/parser-ng-0.1.8.jar  "sin(1,1)"
A SYNTAX ERROR OCCURRED
```
But on the same branch unittest fails. and evaluates it as sin(1*1)

Imust be doing somenting differently now.



No, I had not overloaded nor touched the sin, nor am I plannig to do so, as the issue is not somehow critical (you will type `sin(11)`  isnted of `sin(1+1)`  or  similalry only by typo - thats how I hit it. 
I will split the sinSum PR to two, so that is tracked separately.

Anywaym will split the reprodcur tests - the issue remians here, and will create new issue - as it seesm not related. Sorry for confusion
 (ps. The wget testing of older version, IIRC is what I was doing before)
(pps.  The same program faisl in unit test and ass in cmdline... Will elabroate. No reason for yo to waste time)

**Comment by judovana** [comment]
Hmm. You r "fixes done" had broken a lot of tests.  Even wihtout workaround removed, the tests you had broken shoudl be passing.

The main method runers should be moved to unittets, but I can do it for you later.

**Comment by judovana** [comment]
Looking to the unhappy manual merge in the middle, I would advise to revert the last three commits,  and do PR based on latest master.
Even you yourslef merge the PR right away, at elast the gihub actions may catch what you could have overlooked.

Looking to the fix - it is as scary as I was afradid of. You are not afraid you had caused severe regression?

**Comment by judovana** [comment]
Unluckily, yes. Your fix is very worng:
```
Error:  Failures: 
Error:    MathExpressionTest.customEmbeddedFunctionTest:37 expected: <2> but was: <0>
Error:    MathExpressionTest.customEmbeddedFunctionTestMultipleBrackets:43 expected: <2> but was: <0>
Error:    AlgebraExpressionParserTest.evaluateTest:19 expected: <0> but was: <1>
Error:    DeclarationsTest.hardcodedFunctionCanBeOverwritten:22 expected: <SYNTAX ERROR> but was: <2.0>
Error:    AvgNTest.decimalsInCountWorksAsHould:31 expected: <5> but was: <0>
Error:    AvgNTest.limitWorks:39 expected: <50> but was: <0>
Error:    AvgNTest.testExpres:71 expected: <1> but was: <0>
Error:    AvgNTest.zeroZero:20 expected: <1> but was: <0>
Error:    GeomNTest.decimalsInCountWorksAsHould:32 expected: <3.245342223> but was: <0>
Error:    GeomNTest.limitWorks:40 expected: <4.000000000> but was: <0>
Error:    GeomNTest.testExpres:70 expected: <1> but was: <0>
Error:    GeomNTest.zeroZero:21 expected: <1> but was: <0>
Error:    GsumAvgGeomCountTest.avg:25 expected: <2> but was: <13>
Error:    GsumAvgGeomCountTest.countExpression:54 expected: <5> but was: <1>
Error:    GsumAvgGeomCountTest.geom:33 expected: <2.000000000> but was: <124>
Error:    GsumAvgGeomCountTest.geomExpression:47 expected: <4.000000000> but was: <284>
Error:    GsumAvgGeomCountTest.geomExpressionMultipleBrackets:61 expected: <4.000000000> but was: <64.0>
Error:    GsumAvgGeomCountTest.gsumExpression:41 expected: <6> but was: <1123>
Error:    SumTest.sumAcceptsTwoPlainArguments:28 expected: <11> but was: <56>
```

Please revert it. It caused more damage then it fixed. it is touchig veryu fragile internals. It have to be done properly.

Plainly spoken, now it behaves like `fce(2,8,4)` pass inisde single 284 ....

**Comment by gbenroscience** [comment]
Sorry I don't understand.
fce(2,8,4), what is that?
Things works as expected here

**Comment by gbenroscience** [comment]
I just created a [branch](https://github.com/gbenroscience/ParserNG/tree/fix-dataset-error). The fix is on it. Can you check it out?

**Comment by judovana** [comment]
I had jsut rebased all my  PRs. It is faileng everywhere. Your "fix" hada managed that any function with multipleparams, is not not getting the params.
`anyFunction(1,2,3)` was before your "fix" given list of three number 1,2 and 3. Now it get list of one number - 123.

yes, you fixed thereported  issue indeed,but caused much more severe one.
It cannotbe working. Jsut run unittests. All is failing.

> I just created a [branch](https://github.com/gbenroscience/ParserNG/tree/fix-dataset-error). The fix is on it. Can you check it out?

Please do a PR from that branch. So you can see the unittests. Maybe the isseu is isolated only to extracted functions (which have of course highest coverage, as they were touched). But I do not think so.
If so, then consequent action should be only in move af calls, but it really do nto seem to be that. The extracted functions deals with same set of tokens as hardcoded ones:(

**Comment by judovana** [comment]
Mybe you od not need to - the githuba ctions run also over pushes to main: 
https://github.com/gbenroscience/ParserNG/actions

See:  last passed run; https://github.com/gbenroscience/ParserNG/actions/runs/4066860861
And after your "fixes done":  https://github.com/gbenroscience/ParserNG/actions/runs/4073688611

How could you push withotu even attempting to run the unittetst?

**Comment by judovana** [comment]
```
  @Test
    void prodAcceptsTwoPlainArguments() {
        MathExpression me;
        me = new MathExpression("prod(5,6)");
        Assertions.assertEquals("30.0", me.solve());
    }
```
This is immediate direct regression iin old code, I have never touched:
```
org.opentest4j.AssertionFailedError: 
Expected :30.0
Actual   :56.0
<Click to see difference>

...
	at parser.methods.ProdTest.prodAcceptsTwoPlainArguments(ProdTest.java:29)
...
```

**Comment by judovana** [comment]
thanx a lot for reverting. It is worthy. Will rebase PRs again :)
But double the thanx for attempt  fixing the issue. The fix was close. Just not complet. I had something similar before I worte the workaround, but it is really themost important, most fragile, and most powerfull part of parserng. Crossign fingers!

**Comment by judovana** [comment]
Side note, with the workaround in , my rebased PRs:
https://github.com/gbenroscience/ParserNG/pull/27
https://github.com/gbenroscience/ParserNG/pull/22
https://github.com/gbenroscience/ParserNG/pull/21

are ok to be merged. They all reside out of main development area.

**Comment by judovana** [comment]
Update on `sin`  (but also cosin.. and so on...) issue
`sin(1,1)` correctly exits with SYNTAX error. However `sin(1+1, 2+2)` evaluates as `sin((1+1)*(2+2))`
Tahts why I originally sumitted it as same issue as  the one with `statistical functions`. Se ethe tests in https://github.com/gbenroscience/ParserNG/pull/23 ; they are doin exzactly this.
The issue is indeed reproducible to the uttermost history of parserng

**Comment by judovana** [comment]
```
 $  wget https://repo1.maven.org/maven2/com/github/gbenroscience/parser-ng/0.1.0/parser-ng-0.1.0.jar
...
  $ jshell --class-path parser-ng-0.1.0.jar 
|  Welcome to JShell -- Version 11.0.17
|  For an introduction type: /help intro

jshell> new parser.MathExpression("sin((1+1)*(1+1))").solve()
$1 ==> "-0.7568024953079282"
jshell> new parser.MathExpression("sin(1,1)").solve()
$2 ==> "A SYNTAX ERROR OCCURED"
...
jshell> new parser.MathExpression("sin(4)").solve()
$4 ==> "-0.7568024953079282"
jshell> new parser.MathExpression("sin(1+1,1+1)").solve()
$5 ==> "-0.7568024953079282"

```

**Comment by judovana** [comment]
But I definitly agree that if it is not the same issue, it is ok to solve it separately, and with minimal priority.

**Comment by gbenroscience** [comment]
Sorry, I traced the issue about 2 hours ago, but I was busy.
Once the fix is done, I will update again.

**Comment by gbenroscience** [comment]
Alright, please check PR #28

**Comment by gbenroscience** [comment]
@judovana , still waiting...

**Comment by judovana** [comment]
I had just arrived. Am on it.

**Comment by gbenroscience** [comment]
Closing!

**Comment by judovana** [comment]
Right. ty!
